// @backstage/plugin-gcalendar/dist/index.esm.js
import { createApiRef, createRouteRef, createPlugin, createApiFactory, googleAuthApiRef, fetchApiRef, createComponentExtension } from '@backstage/core-plugin-api';
import { ResponseError } from '@backstage/errors';

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
const gcalendarApiRef = createApiRef({
  id: "plugin.gcalendar.service"
});
class GCalendarApiClient {
  constructor(options) {
    __publicField(this, "authApi");
    __publicField(this, "fetchApi");
    this.authApi = options.authApi;
    this.fetchApi = options.fetchApi;
  }
  async get(path, params) {
    const query = new URLSearchParams(params);
    const url = new URL(
      `${path}?${query.toString()}`,
      "https://www.googleapis.com"
    );
    const token = await this.authApi.getAccessToken();
    const response = await this.fetchApi.fetch(url.toString(), {
      headers: token ? { Authorization: `Bearer ${token}` } : {}
    });
    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }
    return response.json();
  }
  async getCalendars(params) {
    return this.get(
      "/calendar/v3/users/me/calendarList",
      params
    );
  }
  async getEvents(calendarId, params) {
    return this.get(
      `/calendar/v3/calendars/${encodeURIComponent(calendarId)}/events`,
      params
    );
  }
}

var ResponseStatus = /* @__PURE__ */ ((ResponseStatus2) => {
  ResponseStatus2["needsAction"] = "needsAction";
  ResponseStatus2["accepted"] = "accepted";
  ResponseStatus2["declined"] = "declined";
  ResponseStatus2["maybe"] = "tentative";
  return ResponseStatus2;
})(ResponseStatus || {});

const rootRouteRef = createRouteRef({
  id: "gcalendar"
});

const gcalendarPlugin = createPlugin({
  id: "gcalendar",
  routes: {
    root: rootRouteRef
  },
  apis: [
    createApiFactory({
      api: gcalendarApiRef,
      deps: { authApi: googleAuthApiRef, fetchApi: fetchApiRef },
      factory(deps) {
        return new GCalendarApiClient(deps);
      }
    })
  ]
});
const HomePageCalendar = gcalendarPlugin.provide(
  createComponentExtension({
    name: "HomePageCalendar",
    component: {
      lazy: () => import('./esm/index-305a517e.esm.js').then((m) => m.HomePageCalendar)
    }
  })
);

export { GCalendarApiClient, HomePageCalendar, ResponseStatus, gcalendarApiRef, gcalendarPlugin };
